messages/dotnet/messages.proto

```diff
@@ -68,6 +68,7 @@ message Envelope {
     StepDefinition step_definition = 25;
     Hook hook = 26;
     ParameterType parameter_type = 27;
+    UndefinedParameterType undefined_parameter_type = 28;
```

Comment:
Implementation wise I would suggest adding a result object to the `TestRunFinished` event with the undefined parameter type exception. That way consumers of the ndjson can easily see why the test run failed.

Reply:
That's assuming I can convince you to go for option 1.

Reply:
There may be multiple undefined parameter types, so if we bundle that into `TestRunFinished` it would have to be a `repeated` field. Maybe that's ok, but we could also send it as individual messages like the current code in this PR does. I don't think that would be a problem.

Reply:
It does make it more complex to know if a run failed or passed. In the absence of any test results consumers would now have to know that undefined parameter type messages are in fact failures and not merely a special kind of step definition.

I would also have good use case for a more generic result field in `TestRunFinished`. Undefined parameter types are but one reason why Cucumber can fail to startup. For example the DI container may fail to initialize, invalid regexes, errors while instantiating the glue, ect.

Reply:
I could also use the `result`  field in `TestRunFinished` to store the results of `BeforeAll` and `AfterAll` hooks should they fail.

Reply:
The front-end should print snippets for all undefined parameter types. In order to do that, we'd have to embed all the `UndefinedParameterType` messages in `TestRunFinished`.

As you point out there can be many reasons a test run fails, and we need to be able to distinguish between them. How about this:

```protobuf
message TestRunFinished {
  bool success = 1;
  Timestamp timestamp = 2;
  // Error message. Can be a stack trace from a failed `BeforeAll` or `AfterAll`.
  // If there are undefined parameter types, the message is simply "There were undefined parameter types", and the `undefinedParameterTypes` field will have the details
  string message = 3;
  repeated UndefinedParameterType undefinedParameterTypes = 4;
}
```

It's tempting to embed `TestResult` in `TestRunFinished`, but it's really a `TestStepResult`, so I think we should rename it.

WDYT?

Reply:
How about: 

1. Undefined Parameter Types should fail the build. So `TestRunFinished.success = false` and include a message about the missing parameter type.

2. Then we define `TestRunFinished`:

```
message TestRunFinished {
  bool success = 1;
  Timestamp timestamp = 2;
  // Error message. Can be a stack trace from a failed `BeforeAll` or `AfterAll`.
  // If there are undefined parameter types, the message is simply "There were undefined parameter types", and the `undefinedParameterTypes` field will have the details
  string message = 3;
}
```

3. `UndefinedParameterType` as a message for specialized reporters that print snippets.

Reply:
1. The error message should include a message about the undefined parameter type*s* (plural)

2. Agree!
3. Agree!

So if I simply add `message` to `TestRunFinished` we can merge this, right?

Reply:
Yes! That is workable! Please make it a new major version too.

Reply:
Will do - and I'll rename `TestResult` to `TestStepResult` as well, because that's what it is.